File: src/test/java/neoflix/_14_PersonListTest.java
// Task: Rewrite the AuthService to allow users to authenticate against the database
// Outcome: A user will be able to authenticate against their database record
package neoflix;

import neoflix.services.PeopleService;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.neo4j.driver.Driver;

import static neoflix.Params.Sort.name;
import static org.junit.jupiter.api.Assertions.*;

class _14_PersonListTest {
    private static Driver driver;

    @BeforeAll
    static void initDriver() {
        AppUtils.loadProperties();
        driver = AppUtils.initDriver();
    }

    @AfterAll
    static void closeDriver() {
        if (driver != null) driver.close();
    }

    @Test
    void getPaginatedPersonList() {
        PeopleService peopleService = new PeopleService(driver);

        var limit = 3;

        var output = peopleService.all(new Params(null, name, Params.Order.ASC, limit, 0));
        assertNotNull(output);
        assertEquals(limit, output.size());
        assertEquals(" Aaron Woodley", output.get(0).get("name"));

        var paginated = peopleService.all(new Params(null, name, Params.Order.ASC, limit, limit));
        assertNotNull(paginated);
        assertEquals(limit, paginated.size());
        assertNotEquals(output, paginated);
        assertEquals(" Alejandro González Iñárritu", paginated.get(0).get("name"));
    }

    @Test
    void getOrderedPaginatedPersonList() {
        PeopleService peopleService = new PeopleService(driver);

        var q = "A";

        var first = peopleService.all(new Params(q, name, Params.Order.ASC, 1, 0));
        var last = peopleService.all(new Params(q, name, Params.Order.DESC, 1, 0));
        assertNotNull(first);
        assertEquals(1, first.size());
        assertEquals(" Aaron Woodley", first.get(0).get("name"));
        assertNotEquals(first, last);
        assertEquals("Álex Angulo", last.get(0).get("name"));
    }

    @Test
    void getOrderedPaginatedQueryForPersons() {
        PeopleService peopleService = new PeopleService(driver);

        var first = peopleService.all(new Params(null, name, Params.Order.ASC, 1, 0));
        assertNotNull(first);
        assertEquals(1, first.size());

        System.out.println("""
                
                Here is the answer to the quiz question on the lesson:
                What is the name of the first person in the database in alphabetical order?
                Copy and paste the following answer into the text box:
                """);
        System.out.println(first.get(0).get("name").toString().trim());
    }
}

File: pkg/challenges/14_person_list_test.go
package challenges_test

import (
	"fmt"
	"github.com/neo4j-graphacademy/neoflix/pkg/fixtures"
	"testing"

	"github.com/neo4j-graphacademy/neoflix/pkg/config"
	"github.com/neo4j-graphacademy/neoflix/pkg/routes/paging"
	"github.com/neo4j-graphacademy/neoflix/pkg/services"
)

func TestPersonList(t *testing.T) {
	// Load Settings
	settings, err := config.ReadConfig("../../config.json")
	assertNilError(t, err)

	// Init Driver
	driver, err := config.NewDriver(settings)
	assertNilError(t, err)

	defer func() {
		assertNilError(t, driver.Close())
	}()

	service := services.NewPeopleService(
		&fixtures.FixtureLoader{Prefix: "../.."},
		driver)
	assertNotNil(t, service)

	// retrieve a paginated list people from the database
	limit := 10

	output, err := service.FindAll(paging.NewPaging("", "name", "asc", 0, limit))

	assertNilError(t, err)
	assertNotNil(t, output)
	assertEquals(t, limit, len(output))

	paginated, err := service.FindAll(paging.NewPaging("", "name", "asc", limit, limit))

	assertNilError(t, err)
	assertNotNil(t, paginated)
	assertEquals(t, limit, len(paginated))

	assertNotEquals(t, output[0]["name"], paginated[0]["name"])

	// apply a filter, ordering and pagination to the query
	q := "A"

	filteredFirst, err := service.FindAll(paging.NewPaging(q, "name", "asc", 0, 1))

	assertNilError(t, err)
	assertNotNil(t, filteredFirst)
	assertEquals(t, 1, len(filteredFirst))

	filteredLast, err := service.FindAll(paging.NewPaging(q, "name", "desc", 0, 1))

	assertNilError(t, err)
	assertNotNil(t, filteredLast)
	assertEquals(t, 1, len(filteredLast))

	assertNotEquals(t, filteredLast[0]["name"], filteredFirst[0]["name"])

	// Quiz answer
	fmt.Println()
	fmt.Println()
	fmt.Println("Here is the answer to the quiz question on the lesson:")
	fmt.Println("What is the name of the first person in the database in alphabetical order?")
	fmt.Println("Copy and paste the following answer into the text box:")

	fmt.Println(output[0]["name"])
}
